Sources/SignalRClient/HubConnection.swift

```diff
@@ -432,6 +441,56 @@ public class HubConnection {
     fileprivate func connectionDidReconnect() {
         initiateHandshake()
     }
+
+    private func resetKeepAlive() {
+        if connection.inherentKeepAlive {
+            return
+        }
+
+        logger.log(logLevel: .debug, message: "Reset keep alive")
+        keepAlivePingTask?.cancel()
+
+        keepAlivePingTask = DispatchWorkItem { self.sendKeepAlivePing() }
+
+        DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + keepAliveIntervalInSeconds) {
+            if let keepAlivePingTask = self.keepAlivePingTask {
+                keepAlivePingTask.perform()
+            }
+        }
+    }
+
+    private func sendKeepAlivePing() {
+        logger.log(logLevel: .debug, message: "Send keep alive called")
+        guard handshakeStatus.isHandled else {
+            logger.log(logLevel: .debug, message: "Send keep alive called but not connected")
+            keepAlivePingTask = nil
+            return
+        }
+
+        do {
+            let cachedPingMessage = try hubProtocol.writeMessage(message: PingMessage.instance)
+            logger.log(logLevel: .debug, message: "Send keep alive")
+            connection.send(data: cachedPingMessage, sendDidComplete: { error in
+                if let error = error {
+                    self.logger.log(logLevel: .error, message: "Keep alive Error recevied \(error.localizedDescription)")
+                    self.keepAlivePingTask = nil
```

Comment:
If this error occurs what happens to further keep-alives? Looks like we would lose them. Should we actually keep trying to send them and rely on clean up to break the cycle?

Reply:
Good point I changed it.